### PR TITLE
[9076] Remove CE chunked header from t.w.http CONNECT response

### PR DESCRIPTION
**Pull request by izzy**
According to [RFC 7231 Section 4.3.6](https://tools.ietf.org/html/rfc7231#page-31) "A server MUST NOT send any Transfer-Encoding or Content-Length header fields in a 2xx (Successful) response to CONNECT"

Fix for [Trac ticket 9076](https://twistedmatrix.com/trac/ticket/9076#ticket)

### CODE REVIEW

**Comment by exarkun** [comment]
Thanks.  Please add test coverage and a news fragment.

**Comment by izzy** [comment]
Are there any resources on how to add those tests? I'm new to developing twisted itself and kinda lost here what to do and how to do it. Especially got a problem with the fact that twisted does not fully support CONNECT requests. Any input on how to get started with this?

**Comment by izzy** [comment]
It seems that codecov report is a "little" off? Well anyway, added a test, added the topfile, as far as I can tell that should be it.

**Comment by exarkun** [comment]
> It seems that codecov report is a "little" off? Well anyway, added a test, added the topfile, as far as I can tell that should be it.

Yea, the codecov reports aren't really useful because there is so much random variation.  Sorry about that.  Thanks for adding the test - sorry I didn't get a chance to give you any hints on that.

**Comment by glyph** [comment]
While codecov definitely has accuracy issues, in _this_ case it's just reporting a lower number because it reports combined coverage, and no project member has pushed this change to a branch so the buildbots can run the full suite yet.  (This is also why the OS X builder is showing an unfinished status).

File: src/twisted/web/http.py [hunk 1]
```diff
@@ -1035,6 +1039,10 @@ def write(self, data):
                             category=DeprecationWarning, stacklevel=2)
                         # Backward compatible cast for non-bytes values
                         value = networkString('%s' % (value,))
+                    if self.method == b"CONNECT" and name == b"Content-Length":
+                        log.msg("Warning: Removing Content-Length header in"
+                                " response to CONNECT request")
```
Comment (njsmith) [inline src/twisted/web/http.py]: `Content-Length` is totally unobjectionable in, say, a 4xx response to a CONNECT. It's only 2xx responses that are special.

File: src/twisted/web/http.py [hunk 2]
```diff
@@ -1008,9 +1008,13 @@ def write(self, data):
...
             # chunked mode, so that we can support pipelining in
             # persistent connections.
+            # For no-body responses and CONNECT requests this
+            # does not apply and the header should not or must
+            # not be sent.
             if ((version == b"HTTP/1.1") and
                 (self.responseHeaders.getRawHeaders(b'content-length') is None) and
-                self.method != b"HEAD" and self.code not in NO_BODY_CODES):
+                self.method != b"HEAD" and self.code not in NO_BODY_CODES and
+                (self.method != b"CONNECT" and self.code/100 == 2)):
```
Comment (njsmith) [inline src/twisted/web/http.py]: Note that you need `//` here on python 3.

**Comment by wsanchez** [comment]
Test pass except for missing news fragment.  Instructions for those are here: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

**Comment by izzy** [comment]
@wsanchez I added already one.. not sure why that's triggering as error...?

**Comment by rodrigc** [comment]
@mkzero you submitted your change earlier.  After you submitted your change, unfortunately the process for newsfiles changes, *after* you submitted your change.
What you need to do is **git mv** one of your files from:
**src/twisted/web/topfiles/9076.bugfix** to **src/twisted/web/newsfragments/9076.bugfix**.

After you do that, commit, and push the change.

**Comment by rodrigc** [comment]
Also, before  you push your change to this branch, can you merge latest trunk into your branch?
There have been a lot of changes.

**Comment by izzy** [comment]
So I fixed a few things merged trunk into my branch, but I'm not sure if the error remaining in [Travis](https://travis-ci.org/twisted/twisted/jobs/283736825#L3153) is my fault..?

**Comment by rodrigc** [comment]
It's not your fault.  Travis started using Python 3.6.3, and something changed there causing test_sslverify.py to break

**Comment by izzy** [comment]
@rodrigc so AFAICS my PR should be mergeable?

**Comment by rodrigc** [comment]
This cannot be merged, because the failing Travis build will block this.
This is affecting all PR's, not just yours.

@mkzero Can you code review this: by clicking the green *Review changes* button at https://github.com/twisted/twisted/pull/910/files 

I need someone to code review that so I can merge it into trunk.  If I can get this fix into trunk, we can merge trunk into your branch, and get your branch passing Travis again.

**Comment by rodrigc** [comment]
@mkzero I got someone else to code review https://github.com/twisted/twisted/pull/910 and it has been merged to trunk.  Can you merge Twisted trunk into your branch?  That will get Travis passing on your branch.

**Comment by izzy** [comment]
Merged trunk into my branch but it seems the buildbot check hangs?

**Comment by rodrigc** [comment]
That's normal, since only a Twisted committer can trigger all the buildbots.

File: src/twisted/web/test/test_http.py [hunk 3]
```diff
@@ -3213,6 +3213,35 @@ def test_expect100ContinueHeader(self):
...
 
 
+class ConnectRequestTests(unittest.TestCase, ResponseTestMixin):
+    """
+    Test correct RFC 7231 handling for CONNECT requests. Responses to these
+    tests must not contain Transfer-Encoding or Content-Length headers
+    """
+
+    def test_headers(self):
+        transport = StringTransport()
```
Comment (rodrigc) [inline src/twisted/web/test/test_http.py]: Please add docstring here to eliminate error here: https://travis-ci.org/twisted/twisted/jobs/284299844.
When writing docstring, follow this guideline: https://jml.io/pages/test-docstrings.html
and don't put **Test** in the description.

**Comment by izzy** [comment]
Seems like the tests finally are all green. Anything else holding it back from being merged into trunk?

**Comment by Aurora-Beta** [comment]
I'm affected by the issue as well. Can anyone say when i can expect this to be merged?

**Comment by izzy** [comment]
Can someone trigger the buildbot and merge this if possible? I've been waiting almost a year to get this bugfix into Twisted now...

**Comment by exarkun** [comment]
Triggered buildbot

**Comment by exarkun** [comment]
The fedeora24-py3.5 failure is unrelated to this PR so it shouldn't block the merge.  It's the known issue - https://twistedmatrix.com/trac/ticket/9382

File: src/twisted/web/test/test_http.py [hunk 4]
```diff
@@ -3325,6 +3325,41 @@ def test_expect100ContinueHeader(self):
...
+        channel.requestFactory = DummyHTTPHandlerProxy
+        channel.makeConnection(transport)
+        channel.dataReceived(b"CONNECT example.org:443 HTTP/1.1\r\n")
+        channel.dataReceived(b"Host: example.org:443\r\n")
+
+        self.assertEqual(transport.value(), b"")
+        channel.dataReceived(b"\r\n")
+
+        self.assertTrue(
+            transport.value().startswith(b"HTTP/1.1 200 OK\r\n"))
```
Comment (exarkun) [inline src/twisted/web/test/test_http.py]: This assertion looks redundant with the following.  Is there a particular reason to have it?  If not, I think it should be removed.  The `assertResponseEquals` will cover this check.

File: src/twisted/web/test/test_http.py [hunk 5]
```diff
@@ -3325,6 +3325,41 @@ def test_expect100ContinueHeader(self):
 
 
 
+class ConnectRequestTests(unittest.TestCase, ResponseTestMixin):
+    """
+    Test correct RFC 7231 handling for CONNECT requests. Responses to these
```
Comment (exarkun) [inline src/twisted/web/test/test_http.py]: You can leave off "correct" here.  Tests are always for correct behavior. :wink:

File: src/twisted/web/test/test_http.py [hunk 6]
```diff
@@ -3325,6 +3325,41 @@ def test_expect100ContinueHeader(self):
...
+    tests must not contain Transfer-Encoding or Content-Length headers
+    """
+
+    def test_headers(self):
+        """
+        A HTTP connection is opened and a CONNECT request is received, which
```
Comment (exarkun) [inline src/twisted/web/test/test_http.py]: `An HTTP ...`

File: src/twisted/web/test/test_http.py [hunk 7]
```diff
@@ -3325,6 +3325,41 @@ def test_expect100ContinueHeader(self):
...
+        should return a 200 response with the HTTP version header, the host and
+        port as well as the CONNECT command but not a Transfer-Encoding or
+        Content-Length header, which both are not allowed in a 200 response.
```
Comment (exarkun) [inline src/twisted/web/test/test_http.py]: `... neither of which is allowed in a 200 response.` would be more natural English.

File: src/twisted/web/http.py [hunk 8]
```diff
@@ -1085,6 +1089,11 @@ def write(self, data):
                             category=DeprecationWarning, stacklevel=2)
                         # Backward compatible cast for non-bytes values
                         value = networkString('%s' % (value,))
+                    if (self.method == b"CONNECT" and self.code//100 == 2 and
```
Comment (exarkun) [inline src/twisted/web/http.py]: There should be space surrounding the `//` operator.

File: src/twisted/web/http.py [hunk 9]
```diff
@@ -1056,9 +1056,13 @@ def write(self, data):
...
             # chunked mode, so that we can support pipelining in
             # persistent connections.
+            # For no-body responses and CONNECT requests this
+            # does not apply and the header should not or must
+            # not be sent.
             if ((version == b"HTTP/1.1") and
                 (self.responseHeaders.getRawHeaders(b'content-length') is None) and
-                self.method != b"HEAD" and self.code not in NO_BODY_CODES):
+                self.method != b"HEAD" and self.code not in NO_BODY_CODES and
+                (self.method != b"CONNECT" and self.code//100 == 2)):
```
Comment (exarkun) [inline src/twisted/web/http.py]: There should be space surrounding the `//` operator.

File: src/twisted/web/newsfragments/9076.bugfix [hunk 10]
```diff
@@ -0,0 +1 @@
+twisted.web.http now handles CONNECT requests according to RFC 7231 and will not send Content-Length or Transfer-Encoding headers anymore
```
Comment (exarkun) [inline src/twisted/web/newsfragments/9076.bugfix]: Please add a full stop (`.`) to the end of this sentence.

File: src/twisted/web/http.py [hunk 11]
```diff
@@ -1085,6 +1089,11 @@ def write(self, data):
                             category=DeprecationWarning, stacklevel=2)
                         # Backward compatible cast for non-bytes values
                         value = networkString('%s' % (value,))
+                    if (self.method == b"CONNECT" and self.code//100 == 2 and
+                        name == b"Content-Length"):
+                        log.msg("Warning: Removing Content-Length header in"
```
Comment (exarkun) [inline src/twisted/web/http.py]: It looks like the logging style in this module has changed in the interim.  Can you make this `self._log.info(...)` instead (I would almost say something about not putting "Warning" there but I see that other code in this module emits info messages that start with "Warning" so ...)

File: src/twisted/web/http.py [hunk 12]
```diff
@@ -1056,9 +1056,13 @@ def write(self, data):
             # if we don't have a content length, we send data in
             # chunked mode, so that we can support pipelining in
             # persistent connections.
+            # For no-body responses and CONNECT requests this
+            # does not apply and the header should not or must
```
Comment (exarkun) [inline src/twisted/web/http.py]: is it "*should* not" or "*must* not"?  seems like it should be one or the other.

**Review by exarkun — APPROVED** [review]
Thanks.  Sorry about the long review delay.  I've left some minor comments inline.  I'd like to see them addressed but then I think this can be merged.

**Comment by izzy** [comment]
@exarkun I've implemented the changes, tests seem to run fine, can you run the buildbot again?

**Comment by twm** [comment]
I merged this forward in #1358. Closing this one as the original fork has disappeared, so it can't go anywhere.

**Comment by izzy** [comment]
Just for completeness sake, I didn't delete the fork per se, I guess Github just didn't handle me changing my username properly, sorry if that caused any inconveniences.

**Comment by twm** [comment]
@izzy Thanks for clarifying!

**Pull request closed without merging**